### PR TITLE
fixes #11 : add functionality for compute_metrics to accept list of lists apart from files

### LINKED ISSUE #11: add functionality for compute_metrics to accept list of lists apart from files
extend the OO API

### PR DESCRIPTION
**Pull request by kracwarlock**
fixes #11

### CODE REVIEW

File: nlgeval/__init__.py [hunk 1]
```diff
@@ -168,7 +168,7 @@ def load_glove(self):
         self.np = np
         self.glove_emb = Embedding()
 
-    def evaluate(self, ref, hyp):
```
Comment (juharris) [inline nlgeval/__init__.py]: Can we keep the method name the same to reduce compatibility issues?

Comment (kracwarlock) [reply nlgeval/__init__.py]: @temporaer what do you think?

Comment (kracwarlock) [reply nlgeval/__init__.py]: @juharris i think f67414b is more consistent. just need to update the readme as well.

**Comment by AmitMY** [comment]
Before merging, I tested this branch, and I am getting a weird error.
```
print len(refs), len(candidates)
corpus_metrics = nlgeval.compute_metrics(refs, candidates)
```

Output:
```
891 891
Traceback (most recent call last):
  File "/home/nlp/amit/thesis/eval-submissions.py", line 62, in <module>
    corpus_metrics = nlgeval.compute_metrics(refs, candidates)
  File "/home/nlp/amit/thesis/nlgeval/__init__.py", line 215, in compute_metrics
    assert len(refs) == len(hyps)
AssertionError
```


Hyps and refs in JSON:
[bug.json.zip](https://github.com/Maluuba/nlg-eval/files/1792566/bug.json.zip)

**Comment by temporaer** [comment]
There is a strong assumption in the code that the number of references are the same for each hypothesis. I assume that the `refs` parameter you pass has the same length as `hyps`. It should be the transpose, though: a list of reference sources, each containing one string per string in `hyps`.

**Comment by AmitMY** [comment]
> It should be the transpose, though: a list of reference sources, each containing one string per string in hyps.

I don't understand why 1 string per string in hyp, I thought it's 1 list per 1 string in hyp.

If I try to understand what you are saying in a different way, I can for example, if I have maximum 7 references, instead of using the following is a reference:
```
["a"]
```

I need to use:
```
["a", "a", "a", "a", "a", "a", "a"]
```

So I just duplicate references to the max reference size? (which doesn't change the result, but changes the input)

If that is the case I'll try to just fix it in the repo itself.

**Comment by temporaer** [comment]
Duplicating the string would be an ugly workaround, yes. I agree that fixing the code to provide more flexibility would be a better approach. I think one way to go would be to return numerator and denominator  separately for the corpus metrics. You could then can call `compute_individual_metrics` repeatedly compute return the ratio in the end. I don't think there's currently much benefit from batching, so you wouldn't lose too much time. Maybe @kracwarlock can provide more details.

**Comment by AmitMY** [comment]
But, [here](https://github.com/Maluuba/nlg-eval/pull/12/files#diff-fca582dc125100787146fadf042de07fR215) instead of treating the lists as lists for a corresponding hypothesis, I think you treat them like they are list of first reference, list of second reference etc...

Returning the n and d would be a great solution.

**Comment by temporaer** [comment]
the list is being "transposed" a few lines above:
```python
ref_list = [map(str.strip, refs) for refs in zip(*ref_list)]
```

**Comment by kracwarlock** [comment]
The `README.md` file for this pull request explains the input format. Merging.

**Pull request merged**